network/crl.go

```diff
@@ -0,0 +1,380 @@
+	crlFromCertIgnoreStr = "ignore"
+)
+
+const (
+	crlFromCertUse int = iota
+	crlFromCertIgnore
+)
+
+// CRLConfig contains configuration related to certificate validation using CRL
+type CRLConfig struct {
+	uri      string
+	fromCert int // crlFromCert*
+}
+
+// NewCRLConfig creates new CRLConfig
+func NewCRLConfig(uri, fromCert string) (*CRLConfig, error) {
+	_, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	var fromCertVal int
+	switch fromCert {
+	case crlFromCertUseStr:
+		fromCertVal = crlFromCertUse
+	case crlFromCertIgnoreStr:
+		fromCertVal = crlFromCertIgnore
+	default:
+		return nil, errors.New("Invalid `tls_crl_from_cert` value '" + fromCert + "', should be one of 'use', 'ignore'")
+	}
+
+	if uri != "" {
+		// Since this is CRL configuration alone, we don't have access to cache yet;
+		// so let's just download the CRL and forget about it
+		_, err = DefaultCRLClient.Fetch(DefaultCRLClient{}, uri)
+		if err != nil {
+			log.WithError(err).Warnf("CRL: Cannot fetch configured URI '%s'", uri)
+		}
+	}
+
+	return &CRLConfig{uri: uri, fromCert: fromCertVal}, nil
+}
+
+// CRLClient is used to fetch CRL from some URI
+type CRLClient interface {
+	// Fetch fetches CRL from passed URI (can be either http:// or file://)
+	Fetch(uri string) ([]byte, error)
+}
+
+// DefaultCRLClient is a default implementation of CRLClient
+// (as opposed to stub ones used in tests)
+type DefaultCRLClient struct{}
+
+// Fetch fetches CRL from passed URI (can be either http:// or file://)
+func (c DefaultCRLClient) Fetch(uri string) ([]byte, error) {
+	if strings.HasPrefix(uri, "http://") {
+		httpRequest, err := http.NewRequest(http.MethodGet, uri, nil)
+		if err != nil {
+			return nil, err
+		}
+		crlURL, err := url.Parse(uri)
+		if err != nil {
+			return nil, err
+		}
+		httpRequest.Header.Add("Accept", "application/pkix-crl, application/pem-certificate-chain")
+		httpRequest.Header.Add("host", crlURL.Host)
+		httpClient := &http.Client{}
+		httpResponse, err := httpClient.Do(httpRequest)
+		if err != nil {
+			return nil, err
+		}
+		defer httpResponse.Body.Close()
+		content, err := ioutil.ReadAll(httpResponse.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		if httpResponse.StatusCode != 200 {
+			return nil, fmt.Errorf("Server returned non-OK status: %s", httpResponse.Status)
+		}
+
+		return content, nil
+	} else if strings.HasPrefix(uri, "file://") {
+		content, err := ioutil.ReadFile(uri[7:])
+		if err != nil {
+			return nil, err
+		}
+
+		return content, nil
+	}
+
+	return nil, fmt.Errorf("Cannot fetch CRL from '%s', unsupported protocol", uri)
+}
+
+// CRLCache is used to store fetched CRLs to avoid downloading the same URI more than once
+type CRLCache interface {
+	Get(key string) ([]byte, error)
+	Put(key string, value []byte) error
+	Remove(key string) error
+}
+
+// DefaultCRLCache is a default implementation of CRLCache, internally stores items in sync.Map
+type DefaultCRLCache struct {
```

Comment:
can we reuse cache from keystore to not introduce one more and maintain less?
https://github.com/cossacklabs/acra/blob/master/keystore/lru/cache.go - they have similar interfaces on 50% )
plus I don't see where we actually use this DefaultCRLCache

Reply:
> can we reuse cache from keystore to not introduce one more and maintain less?

That one was designed to store `[]byte` while we want to store `pkix.CertificateList`. `Get()` tries to cast value to `[]byte` before returning it. Also, calling `NewCacheKeystoreWrapper()` in context of CRL sounds confusing. Much easier to copy the implementation and change it a little.

>  plus I don't see where we actually use this DefaultCRLCache

It is used to store fetched CRLs (as raw byte arrays), while `LRUParsedCRLCache` stores (a limited amount of) parsed and verified CRLs. You've mentioned once that we may want to use some kind of database instead of in-memory `DefaultCRLCache`.